### PR TITLE
chore: migrate build from azure-pipelines to actions

### PR DESCRIPTION
**Pull request by armano2**
This is POC of build migration from azure-pipelines to actions GitHub Workflow Status

----

From my research about github actions:

- higher concurrent job count (20)
- we can setup tasks/jobs for releases to npm and github packages from within github 
- actions allow us to automate tagging/closing stale tickets, and setup automatic responses
- for branch in same repository as PR, it runs action only once for branch and shares it with PR
- in case of error in matrix step (by default) its stops all other steps from matrix, e.g. 
  https://github.com/armano2/node-markdown-spellcheck/runs/364664300
- other big projects that migrated to actions:
  - typescript: https://github.com/microsoft/TypeScript/actions
  - eslint: https://github.com/eslint/eslint/actions
  - prettier: https://github.com/prettier/prettier/actions
  - babel: https://github.com/babel/babel/actions
  - jest: https://github.com/facebook/jest/actions
  - vscode: https://github.com/microsoft/vscode/actions
  - node: https://github.com/nodejs/node/actions

### Examples
- Build: https://github.com/armano2/typescript-eslint/runs/363745143
- PR: https://github.com/armano2/typescript-eslint/pull/8/checks

#### External project test
- PR: https://github.com/armano2/node-markdown-spellcheck/pull/3

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,102 @@
+name: Node CI
+
+on:
+  push:
+    branches:
+      - master
+  pull_request:
+    branches:
+      - master
```
Comment (bradzacher) [inline .github/workflows/ci.yml]: nit - with bigger pieces of work, we often raise two PRs, with the second targeting the first's branch. Can we just open this to all branches?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+          CI: true
+
+      - name: Run unit tests
+        run: yarn test
+
+  publish_canary_version:
+    name: Publish the latest code as a canary version
+    runs-on: ubuntu-latest
+    needs: [primary_code_validation_and_tests, unit_tests_on_other_node_versions]
+    if: github.event_name == 'push'
```
Comment (bradzacher) [inline .github/workflows/ci.yml]: should this also ensure the push is against master, just in case?

Comment (armano2) [reply .github/workflows/ci.yml]: this is already filtered in `on` but if we really want i can add
```suggestion
    if: github.event_name == 'push' & github.ref == 'refs/head/master'
```

**Comment by armano2** [comment]
> Also do you know how easy it is to setup actions for things like locking old, closed issues? I get annoyed when people necro old things, but have been too lazy to investigate a bot 😅

yea i could set this up to, but this can be done in separate PR

about our bot, there few actions that are already prepeared for this for example:
https://github.com/marketplace/actions/first-interaction
```
name: Greetings

on: [pull_request, issues]

jobs:
  greeting:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/first-interaction@v1
      with:
        repo-token: ${{ secrets.GITHUB_TOKEN }}
        issue-message: 'Message that will be displayed on users'' first issue'
        pr-message: 'Message that will be displayed on users'' first pr'
```

**Comment by JamesHenry** [comment]
Err...why? 😅 also we have a slack channel precisely to discuss things like this

**Comment by armano2** [comment]
@JamesHenry i added small description

**Comment by bradzacher** [comment]
one good reason to move to github actions is because it centralises the admin.
Otherwise I see it as "the same as azure pipelines".

Same for the bots, having them as actions would be a win for maintaining (eg I don't know where or how the PR message bot is implemented atm).

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        env:
+          CI: true
+
+      - name: Publish code coverage report
+        uses: codecov/codecov-action@v1
+        with:
+          yml: ./codecov.yml
+          token: ${{ secrets.CODECOV_TOKEN }}
+          flags: unittest
+          name: codecov
```
Comment (bradzacher) [inline .github/workflows/ci.yml]: Question - did we want to split all of this up to give us more granular status checks?
I know that a bit of a pain point for people is just seeing "primary validation" failed, and then having to dig into the logs to see which step failed.

I.e. one job for each of:
- type check
- format check
- lint
- spelling
- test + integration (+ coverage publish) test on latest node

Comment (armano2) [reply .github/workflows/ci.yml]: you actually don't have to as actions are showing it in overview screen

**Comment by armano2** [comment]
as for codecov/codecov-action we should wait for https://github.com/codecov/codecov-action/pull/53 got get merged first as it has same issue with PR ids

**Pull request merged**